components/locid/src/helpers.rs

```diff
@@ -56,15 +56,150 @@ impl<T> From<Vec<T>> for ShortVec<T> {
     }
 }
 
-macro_rules! impl_writeable_for_single_subtag {
-    ($type:tt, $sample:literal) => {
-        impl core::fmt::Display for $type {
+macro_rules! impl_tinystr_subtag {
+    (
+        $(#[$doc:meta])*
+        $name:ident,
+        $full_name:ty,
+        $($macro_name:ident,)?
+        $len_start:literal..=$len_end:literal,
+        $condition:expr,
+        $normalization:expr,
+        $is_normalized:expr,
+        $error:ident,
+        [$good_example:literal $(,$more_good_examples:literal)*],
+        [$bad_example:literal $(, $more_bad_examples:literal)*],
+    ) => {
+        #[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord, Copy)]
+        #[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+        #[repr(transparent)]
+        $(#[$doc])*
+        pub struct $name(tinystr::TinyAsciiStr<$len_end>);
+
+        impl $name {
+            /// A constructor which takes a utf8 slice, parses it and
+            #[doc = concat!("produces a well-formed [`", stringify!($name), "`].")]
+            ///
+            /// # Examples
+            ///
+            /// ```
+            #[doc = concat!("use icu_locid::", stringify!($full_name), ";")]
+            ///
+            #[doc = concat!("let lang = ", stringify!($name), "::from_bytes(b", stringify!($good_example), ")")]
+            /// let lang = Language::from_bytes(b"en")
+            ///     .expect("Parsing failed.");
+            ///
+            #[doc = concat!("assert_eq!(lang, ", stringify!($good_example), ");")]
+            /// assert_eq!(lang, "en");
+            /// ```
+            pub const fn from_bytes(v: &[u8]) -> Result<Self, crate::parser::errors::ParserError> {
+                Self::from_bytes_manual_slice(v, 0, v.len())
+            }
+
+            /// Equivalent to [`from_bytes(bytes[start..end])`](Self::from_bytes),
+            /// but callable in a `const` context (which range indexing is not).
+            pub const fn from_bytes_manual_slice(
+                v: &[u8],
+                start: usize,
+                end: usize,
+            ) -> Result<Self, crate::parser::errors::ParserError> {
+                let slen = end - start;
+
+                #[allow(clippy::double_comparisons)] // if len_start == len_end
+                if slen < $len_start || slen > $len_end {
+                    return Err(crate::parser::errors::ParserError::$error);
+                }
+
+                match tinystr::TinyAsciiStr::from_bytes_manual_slice(v, start, end) {
+                    Ok(s) if s.len() >= $len_start && $condition(&s) => Ok(Self($normalization(s))),
```

Comment:
`question`: do you need the length condition here? You already test if `slen` is in bounds.

Reply:
slen is the slice's length. However that could contain `\0`s, so the tinystr's length could be smaller

Reply:
oh shoot. Good call, and a nuanced gotcha. Add a comment please for future readers!